src/hosting/runTags.ts

```diff
@@ -0,0 +1,183 @@
+import { FirebaseError } from "../error";
+import { flattenArray } from "../functional";
+
+/**
+ * Sentinel to be used when creating an api.Rewrite with the tag option but
+ * you don't yet know the tag. Resolve this tag by passing the rewrite into
+ * setRewriteTags
+ */
+export const TODO_TAG_NAME = "this is an invalid tag name so it cannot be real";
+
+/**
+ * Looks up all valid Hosting tags in this project and removes traffic targets
+ * from passed in services that don't match a valid tag.
+ * This makes no actual server-side changes to these services; you must then
+ * call run.updateService to save these changes. We divide this responsiblity
+ * because we want to possibly insert a new tagged target before saving.
+ */
+export async function gcTagsForServices(project: string, services: run.Service[]): Promise<void> {
+  // region -> service -> tags
+  // We cannot simplify this into a single map because we might be mixing project
+  // id and number.
+  const validTags: Record<string, Record<string, Set<string>>> = {};
+  const sites = await api.listSites(project);
+  const allVersionsNested = await Promise.all(sites.map((site) => api.listVersions(site.name)));
+  const activeVersions = [...flattenArray(allVersionsNested)].filter((version) => {
+    return version.status === "CREATED" || version.status === "FINALIZED";
+  });
+  for (const version of activeVersions) {
+    for (const rewrite of version?.config?.rewrites || []) {
+      if (!("run" in rewrite) || !rewrite.run.tag) {
+        continue;
+      }
+      validTags[rewrite.run.region] = validTags[rewrite.run.region] || {};
+      validTags[rewrite.run.region][rewrite.run.serviceId] =
+        validTags[rewrite.run.region][rewrite.run.serviceId] || new Set<string>();
+      validTags[rewrite.run.region][rewrite.run.serviceId].add(rewrite.run.tag);
+    }
+  }
+
+  // Erase all traffic targets that have an expired tag and no serving percentage
+  for (const service of services) {
+    // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
+    console.log(`gcpIds of metadta ${JSON.stringify(service.metadata)} is ${run.gcpIds(service)}`);
+    const { region, serviceId } = run.gcpIds(service);
+    service.spec.traffic = (service.spec.traffic || [])
+      .map((traffic) => {
+        // If we're serving traffic irrespective of the tag, leave this target
+        if (traffic.percent) {
+          return traffic;
+        }
+        // Only GC targets with tags
+        if (!traffic.tag) {
+          return traffic;
+        }
+        // Only GC targets with tags that look like we added them
+        if (!traffic.tag.startsWith("fh-")) {
+          return traffic;
+        }
+        if (validTags[region]?.[serviceId]?.has(traffic.tag)) {
+          return traffic;
+        }
+        return null;
+      })
+      // Note: the filter command doesn't update the type info to drop null
+      .filter((t) => t !== null) as run.TrafficTarget[];
```

Comment:
Why map _then_ filter? Couldn't this just have been `.filter` that checks the same conditions and returns true (otherwise false)? (I saw this elsewhere too and came back to check - I believe this is the only place where the type doesn't change (map returns the same values) which is why it could be skipped)